### ISSUE TITLE
seek in boot file petite failed

### ISSUE THREAD

**Issue by riastradh** [racket/racket#5122]
**What version of Racket are you using?**
8.15

**What program did you run?**
make

**What should have happened?**
racket should have built

**If you got an error message, please include it here.**
Last few lines of build output:

```
cc -I/usr/include -I/home/riastradh/pkgsrc/current/pkg/include -I/home/riastradh/pkgsrc/current/pkg/include/freetype2 -I/home/riastradh/pkgsrc/current/pkg/include/glib-2.0 -I/home/riastradh/pkgsrc/current/pkg/include/gio-unix-2.0 -I/home/riastradh/pkgsrc/current/pkg/lib/glib-2.0/include -I/home/riastradh/pkgsrc/current/pkg/include/harfbuzz -I/home/riastradh/pkgsrc/current/pkg/include/libdrm  -pthread -Ics/c/rktio/ -Irktio/ -O2 -g -pthread -I/usr/include -I/home/riastradh/pkgsrc/current/pkg/include -I/home/riastradh/pkgsrc/current/pkg/include/freetype2 -I/home/riastradh/pkgsrc/current/pkg/include/glib-2.0 -I/home/riastradh/pkgsrc/current/pkg/include/gio-unix-2.0 -I/home/riastradh/pkgsrc/current/pkg/lib/glib-2.0/include -I/home/riastradh/pkgsrc/current/pkg/include/harfbuzz -I/home/riastradh/pkgsrc/current/pkg/include/libdrm -o cs/c/rktio/rktio_main.o -c rktio/rktio_main.c
ar rc cs/c/rktio/librktio.a cs/c/rktio/rktio_fs.o cs/c/rktio/rktio_fd.o cs/c/rktio/rktio_file.o cs/c/rktio/rktio_poll_set.o cs/c/rktio/rktio_ltps.o cs/c/rktio/rktio_sleep.o cs/c/rktio/rktio_network.o cs/c/rktio/rktio_pipe.o cs/c/rktio/rktio_process.o cs/c/rktio/rktio_signal.o cs/c/rktio/rktio_envvars.o cs/c/rktio/rktio_fs_change.o cs/c/rktio/rktio_flock.o cs/c/rktio/rktio_shellex.o cs/c/rktio/rktio_time.o cs/c/rktio/rktio_cpu.o cs/c/rktio/rktio_syslog.o cs/c/rktio/rktio_convert.o cs/c/rktio/rktio_sha1.o cs/c/rktio/rktio_sha2.o cs/c/rktio/rktio_dll.o cs/c/rktio/rktio_error.o cs/c/rktio/rktio_hash.o cs/c/rktio/rktio_wide.o cs/c/rktio/rktio_console.o cs/c/rktio/rktio_main.o
cc -O2 -g -pthread -I/usr/include -I/home/riastradh/pkgsrc/current/pkg/include -I/home/riastradh/pkgsrc/current/pkg/include/freetype2 -I/home/riastradh/pkgsrc/current/pkg/include/glib-2.0 -I/home/riastradh/pkgsrc/current/pkg/include/gio-unix-2.0 -I/home/riastradh/pkgsrc/current/pkg/lib/glib-2.0/include -I/home/riastradh/pkgsrc/current/pkg/include/harfbuzz -I/home/riastradh/pkgsrc/current/pkg/include/libdrm -I/usr/include -I/home/riastradh/pkgsrc/current/pkg/include -I/home/riastradh/pkgsrc/current/pkg/include/freetype2 -I/home/riastradh/pkgsrc/current/pkg/include/glib-2.0 -I/home/riastradh/pkgsrc/current/pkg/include/gio-unix-2.0 -I/home/riastradh/pkgsrc/current/pkg/lib/glib-2.0/include -I/home/riastradh/pkgsrc/current/pkg/include/harfbuzz -I/home/riastradh/pkgsrc/current/pkg/include/libdrm  -pthread -o cs/c/raw_racketcs cs/c/main.o cs/c/boot.o cs/c/ChezScheme/ta6nb/boot/ta6nb/libkernel.a cs/c/ChezScheme/ta6nb/lz4/lib/liblz4.a cs/c/rktio/librktio.a -L/home/riastradh/pkgsrc/current/pkg/gcc10/lib/gcc/x86_64--netbsd/10.5.0 -Wl,-R/home/riastradh/pkgsrc/current/pkg/gcc10/lib/gcc/x86_64--netbsd/10.5.0 -Wl,-zrelro -L/usr/lib -Wl,-R/usr/lib -L/home/riastradh/pkgsrc/current/pkg/lib -Wl,-R/home/riastradh/pkgsrc/current/pkg/lib -pthread -pthread -rdynamic -lm  -lm -lpthread  -lncurses -lterminfo -lz
/usr/sbin/paxctl +m cs/c/raw_racketcs
cs/c/ChezScheme/ta6nb/bin/ta6nb/scheme -B cs/c/ChezScheme/ta6nb/boot/ta6nb/petite.boot -B cs/c/ChezScheme/ta6nb/boot/ta6nb/scheme.boot --script cs/c/to-vfasl.ss cs/c/ChezScheme/ta6nb/boot/ta6nb/petite.boot cs/c/petite-v.boot
cs/c/ChezScheme/ta6nb/bin/ta6nb/scheme -B cs/c/ChezScheme/ta6nb/boot/ta6nb/petite.boot -B cs/c/ChezScheme/ta6nb/boot/ta6nb/scheme.boot --script cs/c/to-vfasl.ss cs/c/ChezScheme/ta6nb/boot/ta6nb/scheme.boot cs/c/scheme-v.boot petite
cs/c/ChezScheme/ta6nb/bin/ta6nb/scheme -B cs/c/ChezScheme/ta6nb/boot/ta6nb/petite.boot -B cs/c/ChezScheme/ta6nb/boot/ta6nb/scheme.boot --script cs/c/to-vfasl.ss cs/c/racket.boot cs/c/racket-v.boot petite scheme
cs/c/raw_racketcs -O info'@'compiler/cm -l- setup --boot setup-go.rkt cs/c/compiled ignored cs/c/ignored.d cs/c/embed-boot.rkt --target ta6nb cs/c/raw_racketcs cs/c/racketcs cs/c/petite-v.boot cs/c/scheme-v.boot cs/c/racket-v.boot
seek in boot file petite failed
failed
 in build-one
 in loop
 in module->hash
gmake: *** [Makefile:22: all] Error 1
```

**Please include any other relevant details**
NetBSD 9

ktracing the last cs/c/raw_racketcs incantation shows that the file descriptor passed to lseek is -1:

```
 14398  14398 raw_racketcs CALL  lseek(0xffffffff,0,0,0)
 14398  14398 raw_racketcs RET   lseek -1 errno 9 Bad file descriptor
 14398  14398 raw_racketcs CALL  write(2,0x7f7fffd31d90,0x20)
 14398  14398 raw_racketcs GIO   fd 2 wrote 32 bytes
       "seek in boot file petite failed\n"
```

I tried adding `--verbose` to the cs/c/raw_racketcs command line in an attempt to find which file descriptor open path is failing, but it didn't change anything.

There are some very suspicious open calls in the ktrace output:

```
 14398  14398 raw_racketcs CALL  open(0x761a1a4f0000,0,0x1a4f2800)
 14398  14398 raw_racketcs NAMI  "/proc/14398/petite-v.boot"
 14398  14398 raw_racketcs RET   open -1 errno 2 No such file or directory
...
 14398  14398 raw_racketcs CALL  open(0x761a1a4f0020,0,1)
 14398  14398 raw_racketcs NAMI  "/proc/14398/scheme-v.boot"
 14398  14398 raw_racketcs RET   open -1 errno 2 No such file or directory
 14398  14398 raw_racketcs CALL  open(0x761a1a4f0040,0,1)
 14398  14398 raw_racketcs NAMI  "/proc/14398/racket-v.boot"
 14398  14398 raw_racketcs RET   open -1 errno 2 No such file or directory
```

The commits

- https://github.com/racket/racket/commit/b4a2467f2b2d1c03e10790ef59697f3b2351765d Chez Scheme: find boot files relative to the executable by default
- https://github.com/racket/racket/commit/b4efd760d2982349253b99bf185e6896247cbd2b Chez Scheme: Support executable-relative boot files on more platforms

strikes me as likely culprits.

It looks like what's happening is:

1. ChezScheme/c/main.c passes argv[0]="cs/c/raw_racketcs" as execpath to Sregister_boot_executable_relative_file
2. Sregister_boot_executable_relative_file, via find_boot and next_path, passes execpath="cs/c/raw_racketcs" to S_get_process_executable_path
3. S_get_process_executable_path ignores its argument and replaces it by "/proc/curproc/file" (on NetBSD, Minix, Dragonfly, and AIX) before feeding it through realpath to return "/proc/${getpid()}/file"
4. All the subsequent path resolution is relative to "/proc/${getpid()}/" instead of "cs/c/" as it should be.

It seems to me the bug is step (3) -- if get_process_executable_path has a perfectly good exec_file input, it doesn't need to rummage around in /proc (which doesn't always exist) for anything.  And certainly it won't find boot files in /proc.

**Comment by mflatt** [comment]
Thanks for the report! I will look a little more and probably pursue this first at `cisco/ChezScheme`, but it looks like `"/proc/curproc/file"` should instead be `"/proc/curproc/exe"` for NetBSD to implement the intended behavior.

The failure to check an `open` result is probably Racket-specific, though, and I'll also follow up on that.

**Comment by mflatt** [comment]
Thanks for the recommendation. I see that Racket used to work that way on NetBSD (and FreeBSD), but it got lost along the way. I'll restore that approach for NetBSD.

**Comment by mflatt** [comment]
Fixed by 1fe2ae1ba3d93772488abc7354151dbc3124e5eb and 0a52e493333e3e1eb30ab4df58378cbdde31364a.
